subsys/zbus/zbus.c

```diff
@@ -191,21 +196,142 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
 	return last_error;
 }
 
+static inline void _zbus_chan_update_hop(const struct zbus_channel *chan)
+{
+
+	struct zbus_channel_observation *observation;
+	struct zbus_channel_observation_mask *observation_mask;
+
+	int chan_highest_observer_priority = _MIN_THREAD_PRIORITY;
+
+	K_SPINLOCK(&_zbus_lock)
+	{
+
+		for (int16_t i = chan->data->observers_start_idx,
+			     limit = chan->data->observers_end_idx;
+		     i < limit; ++i) {
+			STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+			STRUCT_SECTION_GET(zbus_channel_observation_mask, i, &observation_mask);
+
+			__ASSERT(observation != NULL, "observation must be not NULL");
+
+			const struct zbus_observer *obs = observation->obs;
+
+			if (!obs->data->enabled || observation_mask->enabled) {
+				continue;
+			}
+
+			if (chan_highest_observer_priority > obs->data->priority) {
+				chan_highest_observer_priority = obs->data->priority;
+			}
+		}
+		chan->data->highest_observer_priority = chan_highest_observer_priority;
+	}
+}
+
+static inline void _zbus_update_all_channels_hop(const struct zbus_observer *obs)
+{
+	struct zbus_channel_observation *observation;
+
+	int count;
+
+	STRUCT_SECTION_COUNT(zbus_channel_observation, &count);
+
+	for (int16_t i = 0; i < count; ++i) {
+		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+
+		if (obs != observation->obs) {
+			continue;
+		}
+
+		_zbus_chan_update_hop(observation->chan);
+	}
+}
+
+static inline void _zbus_obs_attach_to_thread(const struct zbus_observer *obs)
+{
+	__ASSERT(obs != NULL, "obs is required");
+
+	int current_thread_priority = k_thread_priority_get(k_current_get());
+
+	if (unlikely(obs->data->priority != current_thread_priority)) {
+		obs->data->priority = current_thread_priority;
+
+		_zbus_update_all_channels_hop(obs);
+	}
+}
+
+static inline int _zbus_chan_lock(const struct zbus_channel *chan, k_timeout_t timeout, int *prio)
+{
+	bool boosting = false;
+
+	if (IS_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST) && !k_is_in_isr()) {
+		*prio = k_thread_priority_get(k_current_get());
+
+		K_SPINLOCK(&_zbus_lock)
+		{
+			if (*prio > chan->data->highest_observer_priority) {
+
+				int new_prio = chan->data->highest_observer_priority - 1;
+
+				new_prio = MAX(new_prio, 0);
+
+				LOG_DBG("Boosting %s thread priority from %d to %d", _THREAD_NAME,
+					*prio, (int)new_prio);
+
+				k_thread_priority_set(k_current_get(), new_prio);
+
+				boosting = true;
+			}
+		}
+	}
+
+	int err = k_sem_take(&chan->data->sem, timeout);
+
+	if (err) {
+		if (boosting) {
+			LOG_ERR("Something went wrong. Restoring %s priority", _THREAD_NAME);
```

Comment:
LOG_ERR() is almost always going to appear in the log stream, and "Something went wrong" isn't informative enough to be worth logging, IMHO. (FWIW: k_sem_take is only going to return non-zero on a timeout, or plausibly something like someone directly waking the thread up or destroying the semaphore).    The function has its own return value, you can just return that and let the caller log whatever the broken state is.